### DISCUSSION TITLE
How to override / dissable or add rules to cssnano-simple which nextjs uses to minify css styles in production mode?

### DISCUSSION THREAD

**Discussion by grzegorzpokorski** [vercel/next.js#44687] [Help] [answered]
I want to use oklch() functional notation expresses a given color in the OKLCH color space in my CSS files. I use additional postcss plugin [@csstools/postcss-oklab-function](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-oklab-function) to provide a fallback styles for browser which doesn't support oklch notation. In develpoment mode, it works fine, but in production mode I lose a fallback styles. Below I present a simple reproduction example:

https://github.com/AdiPol1359/oklab-crash-reproduction

I think it is caused by [`cssnano-simple`](https://github.com/Timer/cssnano-simple) which nextjs uses to minify and optimize CSS code. And my question is: It is possible to override rules / preset [cssnano-preset-simple](https://github.com/Timer/cssnano-preset-simple) which are used by cssnano-simple to disable some rules which wipe a fallback styles from production (minified and optimized by custom cssnano-simple) css?

**Comment by icyJoseph** [comment, marked as answer]
Hi,

Long ago someone had an issue with a CSS transform, going from RGB to HSLA, and while trying to help them out I made this awful hack that overrides the pre-compiled cssnano-simple plugin, and allows you to customise your own: https://github.com/vercel/next.js/discussions/41288#discussioncomment-3838704

Reposting for visibility:

```js
const cssnanoSimple = require("next/dist/compiled/cssnano-simple");
const postcss = require("postcss");
const {
  CssMinimizerPlugin,
} = require("next/dist/build/webpack/plugins/css-minimizer-plugin");

const { sources } = require("next/dist/compiled/webpack/webpack");

class CustomCssMinizer extends CssMinimizerPlugin {
  optimizeAsset(file, asset) {
    const postcssOptions = {
      ...this.options.postcssOptions,
      to: file,
      from: file,
    };

    let input;
    if (postcssOptions.map && asset.sourceAndMap) {
      const { source, map } = asset.sourceAndMap();
      input = source;
      postcssOptions.map.prev = map ? map : false;
    } else {
      input = asset.source();
    }

    return postcss([cssnanoSimple({ colormin: { exclude: false } }, postcss)])
      .process(input, postcssOptions)
      .then((res) => {
        if (res.map) {
          return new sources.SourceMapSource(res.css, file, res.map.toJSON());
        } else {
          return new sources.RawSource(res.css);
        }
      });
  }
}

/** @type {import('next').NextConfig} */
const nextConfig = {
  reactStrictMode: true,
  webpack(config, { nextRuntime }) {
    if (!nextRuntime) {
      const customCSSMinizer = (compiler) => {
        new CustomCssMinizer({
          postcssOptions: {
            map: {
              // `inline: false` generates the source map in a separate file.
              // Otherwise, the CSS file is needlessly large.
              inline: false,
              // `annotation: false` skips appending the `sourceMappingURL`
              // to the end of the CSS file. Webpack already handles this.
              annotation: false,
            },
          },
        }).apply(compiler);
      };

      config.optimization.minimizer[1] = customCSSMinizer;
    }
    return config;
  },
};

module.exports = nextConfig;
```

One huge gotcha is that `config.optimization.minimizer[1] = customCSSMinizer;` is necessary, basically override the already present plugin.

**Reply by grzegorzpokorski** [reply]
@icyJoseph thanks for your help! It seems like that what I was excacly looking for, but in meanwhile I found a simpler solution for my problem. Fallback styles are confusing for cssnano, because it treats them as a duplicated code or omit them when it collapses longhand properties into the shorthand representation. I got around this problem by simply use CSS variables and `@supports` CSS at-rule - example below:
```css
:root{
 --custom-color: rgb(245, 255, 255);
}

@supports (color: color(display-p3 0 0 0)) {
 :root {
  --custom-color: color(display-p3 0.96859 1.00000 1.00000);
 }
}

.foo2 {
  border-color: var(--custom-color);
}
```

 By doing that I no longer need falback styles like this:
```css
.foo2 {
  border-color: rgba(255,0,0); /* fallback */
  border-color: oklch(62.8%, 64.5%, 29.234%);
}
```
so, sssnano has nothing to remove.

Sometimes the ability to override cssnano configuration in nextjs is the only way to resolve some problems with css optimalization, so I'm glad that You share this hack!

**Reply by netdoer-zpete** [reply]
Thank you for posting this code, it saved me a lot of trouble 😁

I was having in issue where this css:
![Pasted image 20250106151504](https://github.com/user-attachments/assets/435b371e-2926-4465-a5b9-04164b201ebd)

was being minimized to this in production, causing smaller text sizes at larger screen widths:
![Pasted image 20250106151425](https://github.com/user-attachments/assets/cbea56ba-4d00-4b72-baa1-4a05360c073b)

In my next.config.mjs, I used `{ mergeRules: { exclude: true } }`, and that fixed my problem. Great work!!

**Comment by philwolstenholme** [comment]
I'd love a better way of doing this. I'd like to manage my own cssnano config via my Postcss file
